File: scala/robot-name/src/main/scala/RobotName.scala
import scala.collection.mutable.HashSet
import scala.util.Random

class NameHistory {
  private val history: HashSet[String] = HashSet.empty[String]

  def store(name: String): String = {
    history.add(name)
    name
  }
  def unavailable(name: String) = history.contains(name)
}

object RobotName {
  private val history: NameHistory = new NameHistory

  def generate: Iterator[String] =
    Iterator
      .continually(generateName)
      .filterNot(history.unavailable)
      .map(history.store)

  private def generateName: String = letters() + digits

  private def digits: Int = (100 + Random.nextInt(999))

  private def letters(str: String = ""): String = {
    str match {
      case s if s.length == 2 => s.toUpperCase
      case s => {
        letters(s + Random.alphanumeric.filter(_.isLetter).head)
      }
    }
  }
}

File: typescript/robot-name/src/RobotName.ts
import { NameHistory } from './NameHistory';

export class RobotName {
  private _history: NameHistory;

  private _letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('')
  private _digits = '0123456789'.split('')
  private _nameSpec = [
    this._letters,
    this._letters,
    this._digits,
    this._digits,
    this._digits,
  ];

  constructor(history: NameHistory) {
    this._history = history;
  }

  public name(): string {
    return this.generate().next().value;
  }

  private *generate(): Generator<string> {
    const name = this.buildName();

    if (this._history.isUnavailable(name)) {
      yield * this.generate()
    }

    this._history.save(name)

    yield name;
  }

  private buildName(): string {
    return this._nameSpec
      .map((s) => s[Math.floor(Math.random() * s.length)])
      .join('');
  }
}

